volumes.cpp

```diff
@@ -1312,7 +1331,249 @@ SurfaceVolume& makeSurface( const SurfaceCard* card, VolumeCache* v){
     
   }
   
-  
 }
 
+SurfaceVolume* FacetSurface( const std::string mnemonic, const std::vector< double > args, int facet ){
+  if( mnemonic == "rcc" ){
+    return rccFacet( args, facet );
+  }
+  else if(mnemonic == "box"){
+    return boxFacet( args, facet );
+  }
+  else if( mnemonic == "rpp"){
+    return rppFacet( args, facet );
+  }
+  else if( mnemonic == "hex" || mnemonic == "rhp" ){
+    return hexFacet( args, facet );
+  }
+  else if( mnemonic == "rec" ){
+    return recFacet( args, facet );
+  }
+  throw std::runtime_error( mnemonic + " does not have macrobody facet support at this time." );
+}    
+
+
+//This function is used to find D in the planar equation Ax + By + Cz - D = 0 when given two vectors.
+int planePoint( std::vector< double > args, int i, bool end ){
+  if( end ){
+    return args.at(i)*( args.at(0) + args.at(i) ) + args.at(i + 1)*( args.at(1) + args.at(i + 1) ) + args.at(i + 2)*( args.at(2) + args.at(i + 2) );
+  }
+  else{
+    return args.at(i)*args.at(0) + args.at(i + 1)*args.at(1) + args.at(i + 2)*args.at(2);
+  }
+}
+
+
+SurfaceVolume* rccFacet( const std::vector< double > args, int facet ){
+  if( facet == 1 ){
+    //cylinder surface
+    return new RccVolume( Vector3d(args), Vector3d(args,3), args.at(6), true );
+  }
+  else if( facet == 2 ){
+    //plane at end of vector
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, true );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else if( facet == 3 ){
+    //plane at start of vector
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, false );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else{
+    throw std::runtime_error( "rcc only has 3 facets" );
+  }
+}
+
+SurfaceVolume* boxFacet( const std::vector< double > args, int facet ){
+  if( facet == 1 ){
```

Comment:
A lot of repeated code here that is mathematically pretty consistent:
* set first args value by (1,2) vs (3,4) vs (5,6)
* set `end` by odd vs even
call the same code for all 6 cases